darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py

```diff
@@ -0,0 +1,444 @@
+    Uses the DXT records to construct individual
+    dataframes for both read and write segments.
+
+    Parameters
+    ----------
+
+    dict_list: a sequence of DXT records, where each record is a
+    Python dictionary with the following keys: 'id', 'rank',
+    'hostname', 'write_count', 'read_count', 'write_segments',
+    and 'read_segments'. The read/write data is stored in
+    ``read_segments`` and ``write_segments``, where each is a
+    ```pd.DataFrame`` containing the following data (columns):
+    'offset', 'length', 'start_time', 'end_time'.
+
+    ops: a sequence of keys designating which Darshan operations
+    to collect data for. Default is ``["read", "write"]``.
+
+    Returns
+    -------
+
+    rd_wr_dfs: dictionary where each key is an operation from the input
+    ``ops`` parameter (i.e. "read", "write") and each value is a
+    ``pd.DataFrame`` object containing all of the read/write events.
+
+    Notes
+    -----
+
+    Used in ``get_single_df_dict()``.
+
+    Examples
+    --------
+
+    ``dict_list`` and ``rd_wr_dfs``
+    generated from ``tests/input/sample-dxt-simple.darshan``:
+
+        dict_list = [
+            {
+                'id': 14388265063268455899,
+                'rank': 0,
+                'hostname': 'sn176.localdomain',
+                'write_count': 1,
+                'read_count': 0,
+                'write_segments':
+                    offset  length  start_time  end_time
+                    0       0      40    0.103379  0.103388,
+                'read_segments':
+                    Empty DataFrame
+                    Columns: []
+                    Index: []
+            },
+            {
+                'id': 9457796068806373448,
+                'rank': 0,
+                'hostname': 'sn176.localdomain',
+                'write_count': 1,
+                'read_count': 0,
+                'write_segments':
+                    offset  length  start_time  end_time
+                    0       0    4000    0.104217  0.104231,
+                'read_segments':
+                    Empty DataFrame
+                    Columns: []
+                    Index: []
+            },
+        ]
+
+        rd_wr_dfs = {
+            'read':
+                Empty DataFrame
+                Columns: []
+                Index: [],
+            'write':
+                length  start_time  end_time  rank
+                0      40    0.103379  0.103388     0
+                1    4000    0.104217  0.104231     0
+        }
+
+    """
+    # columns to drop when accumulating the dataframes.
+    # Currently "offset" data is not utilized
+    drop_columns = ["offset"]
+    # create empty dictionary to store
+    # the concatenated read/write dataframes
+    rd_wr_dfs = {}
+    # iterate over each operation
+    for op_key in ops:
+        # mypy can't tell that these keys are in fact
+        # in the ``SegDict``, so just ignore the type
+        seg_key = op_key + "_segments"  # type: ignore
```

Comment:
I have it set to ignore the type here because `mypy` would raise an error saying the key wasn't one of the keys specified in `SegDict`. This isn't true, but I don't know another way to get around this.